### PR TITLE
count_even_numbers_in_list

### LINKED ISSUE #12: count_even_numbers_in_list
# Count the Even Numbers in a List

Write a function that takes in a list of integers and returns the count of even numbers.

### PR DESCRIPTION
**Pull request by Norbert250**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/count_even_numbers.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on January 4, 2025
+
```
Comment (byabdallah) [inline solutions/count_even_numbers.py:6]: There should be a module docstring in the function file. This existing module doesn't explain anything about the file. Some description of the module should be included, for example:

A module for counting even numbers in a given list.

Module contents:
    - FUNCTION_NAME: WHAT DOES IT DO
   
Note: I do recognize that such a module docstring would be repeating what's mentioned later in the file. However, it's about building a habit of writing such stuff. In modules that contain much sophisticated code, the issue of repetition will be reduced significantly.

File: solutions/count_even_numbers.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        0
+        >>> count_even_numbers([1, 2, 'three', 4.5])  # Non-integer values raise an error
+        Traceback (most recent call last):
+            ...
+        AssertionError: All elements in the list must be integers.
+    """
+    assert all(
+        isinstance(num, int) for num in numbers
+    ), "All elements in the list must be integers."
+    return sum(1 for num in numbers if num % 2 == 0)
```
Comment (byabdallah) [inline solutions/count_even_numbers.py]: First, I believe that line 42 is not appealing when it comes to readability. It has to be modified so it becomes more readable. Also, The main goal was to check the number of even numbers in a list, so, why does the implementation exclude the entire list if it contains non-integers? I believe this can be done in a better way. This better way involves iterating over each element of a list, if that element is integer, then the code has to check whether it's even or odd. Finally, the code returns the number of even numbers regardless of what other types the list may include. This is my suggestion for a better function, however, it's not obligatory. But, whether the programmer chooses to edit the whole code or not, line 42 should be amended for readability.

File: solutions/count_even_numbers.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+Module contents:
+- count_even_numbers: Counts the even integers in a list while ignoring non-integer elements.
+
+Created on January 8, 2025
+
+@author: Norbert Ndayisenga
+"""
+
+from typing import List
```
Comment (byabdallah) [inline solutions/count_even_numbers.py]: What's the point of this import statement?

File: solutions/tests/test_count_even_numbers.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+Test Suite for count_even_numbers.py
+Created on January 8, 2025
+
+Tests the functionality of the count_even_numbers function.
+
+@author: Norbert Ndayisenga
+"""
+
+import unittest
+from solutins.count_even_numbers import count_even_numbers
```
Comment (byabdallah) [inline solutions/tests/test_count_even_numbers.py]: edit this so it becomes "from ..count_even_numbers import count_even_numbers"

**Review by byabdallah — APPROVED** [review]
